### DISCUSSION TITLE
Modify Classifiers to Include Class Weighting?

### DISCUSSION THREAD

**Discussion by jentlapalm** [alan-turing-institute/sktime#3609] [Q&A]
Hello sktime community!

Is there a way for me to modify or add a class_weight parameter to any of the sktime classifiers similar to the class_weight parameter included in many sklearn models?

I'm working on a classification project with imbalanced data (target variable is ~10% of data); the goal is to maximize recall of the target class. 

The sktime classifiers I've tried have high accuracy, but achieve this by predicting nearly everything as part of the majority class.

**Comment by fkiraly** [comment]
Afaik the classification module does not support class weights currently.

You can try undersampling/oversampling though before you apply classifiers.

Perhaps some of the classification aficionados has a better suggestion.

**Comment by sssilvar** [comment]
I'm affraid I haven't seen this functionality in `sktime` yet. You could either resample using a workaround in `imblearn`. Example:

```python
def _balance_classes(
    target: ArrayLike,
    ts: NDArray[np.float_],
) -> tuple[pd.DataFrame, NDArray[np.float_]]:
    """Balances class distribution by under-sampling majority classes."""
    rus = RandomUnderSampler()  # You could replace this by an Oversampler
    _, y_resampled = rus.fit_resample(metadata, target)
    ts_resampled = ts[y_resampled.index]
    metadata_resampled = metadata.loc[y_resampled.index]
    return metadata_resampled, ts_resampled
   ```

This will help you deal with the panel data `sktime` supports (talking about 3D arrays with shape `(n_instances, n_variables, n_timepoints)`. It would be nice to have though a more advanced resampling like the iterative resamplers in `torch` or `keras` but it'd make the library too dependent on those as the workflow `sktime` seems to use is more like in `sklearn` where training data is immutable (Data panel ->  pipeline).

**Reply by fkiraly** [reply]
hm, under/oversampling should not be too difficult to implement as a transformer component.

Integration with `imblearn` might be a nice project, to apply to classifiers, regressors, etc.

Would you like to open an issue with this suggestion, @sssilvar, so we can track it as a feature request?

**Reply by fkiraly** [reply]
also related to the "data augmentation" idea, FYI @benHeid, @astrogilda.

Under/Oversampling has the same signature-type as bootstrap augmenters.

**Reply by sssilvar** [reply]
Working on it !
Started bad though. I've extended it from `BaseTransformer`. I will adapt it.
